docs/site/Context.md

```diff
@@ -229,3 +229,150 @@ class HelloController {
+   listener receives a binding object which only has the binding key populated.
+
+   A workaround is to create the binding first before add it to a context:
+
+   ```ts
+   const ctx = new Context();
+   const binding = Binding.create('foo')
+     .to('foo-value')
+     .tag('foo-tag');
+   ctx.add(binding);
+   ctx.on('bind', binding => {
+     console.log(binding.tagMap); // returns `['foo-tag']`
+   });
+   ```
+
+2. It's hard for event listeners to perform asynchronous operations.
+
+To make it easy to support asynchronous event processing, we introduce
+`ContextObserver` and corresponding APIs on `Context:
+
+1. ContextObserver interface
+
+```ts
+/**
+ * Observers of context bind/unbind events
+ */
+export interface ContextObserver {
+  /**
+   * An optional filter function to match bindings. If not present, the listener
+   * will be notified of all binding events.
+   */
+  filter?: BindingFilter;
+
+  /**
+   * Listen on `bind`, `unbind`, or other events
+   * @param eventType Context event type
+   * @param binding The binding as event source
+   */
+  observe(
+    eventType: ContextEventType,
+    binding: Readonly<Binding<unknown>>,
+    context: Context,
+  ): ValueOrPromise<void>;
+}
+```
+
+2. Context APIs
+
+- `subscribe(observer: ContextObserver)`
+
+  Add a context event observer to the context chain, including its ancestors
+
+- `unsubscribe(observer: ContextObserver)`
+
+  Remove the context event observer from the context chain
+
+- `close()`
+
+  Close the context and release references to other objects in the context
+  chain. Please note observers of a child context are also registered with its
+  parent context. As a result, the `close` method must be called to avoid memory
+  leak if the child context is to be recycled.
+
+To react on context events asynchronously, we need to implement the
+`ContextObserver` interface and register it with the context.
+
+For example:
+
+```ts
+const app = new Context('app');
+server = new Context(app, 'server');
+
+const observer: ContextObserver = {
+  // Only interested in bindings tagged with `foo`
+  filter: binding => binding.tagMap.foo != null,
+
+  observe(event: ContextEventType, binding: Readonly<Binding<unknown>>) {
+    if (event === 'bind') {
+      console.log('bind: %s', binding.key);
+      // ... perform async operation
+    } else if (event === 'unbind') {
+      console.log('unbind: %s', binding.key);
+      // ... perform async operation
+    }
+  },
+};
+
+server.subscribe(observer);
+server
+  .bind('foo-server')
+  .to('foo-value')
+  .tag('foo');
+app
+  .bind('foo-app')
+  .to('foo-value')
+  .tag('foo');
+
+// The following messages will be printed:
+// bind: foo-server
+// bind: foo-app
+```
+
+It's recommended that `ContextObserver` implementations should not throw errors
+in `observe` method. Uncaught errors will be reported as `error` events of the
+context object. If there is no error listener registered, the node process will
+crash.
```

Comment:
Let's add a new heading before this section, e.g. `### Error handling`. That way we can create a URL pointing directly to this part.

Can you also expand this paragraph to describe how are errors affected by context hierarchy (child vs. parent context)? If you implement my proposal to propagate errors from child contexts to the parent context, then we should describe that behavior here.